Issue 273 add roundtrip efficiency to scheduler

Flix6x opened the pull request:
This PR rewrites our generic device scheduler to:
- Deal with asymmetric efficiency losses of individual devices.
- Deal with asymmetric up and down prices for deviating from previous commitments.

Furthermore, it allows round-trip efficiency to be communicated as a new optional field when POSTing UDI Events, with efficiency losses being assigned equally to charging and discharging.

Closes #273.

NB In a premature state, this PR applied round-trip efficiency by scaling charging and discharging prices accordingly. However, this approach is limited to use cases with a single device per EMS. With tariffs modeled as a commitment to zero prosumption, for a single-device EMS, the up price is equal to the consumption price (charging) and the down price is equal to the feed-in price (discharging).

nhoening reviewed (changes requested):
Looks pretty good, besides one potential error and some documentation I'd like.

And one fundamental question: It seems we are going from one `power` variable to `ems_power`, `device_power_down` and `device_power_up`. Does this mean that in principle (although it is usually isn't optimal) we could charge and discharge at the same time?

Also, I have some notes which don't necessarily block this PR, so you could either address them, or we make a new ticket:

For full understanding, I wanted to see how the battery and charging stations differ. I had a tough time really seeing how that is, as battery.py and charging_station.py seem to share a large amount of code. 

My assumption: They both schedule only one sensor (one for a battery, one for a charging station) against SOC targets for lowest costs. The "EMS" is not being optimized, but it actually is the environment of the 1 device (e.g. the grid).

Is that correct?
The comments are sometimes talking about "devices" (plural) so I'm unsure.

A few smaller questions, maybe addressable in this PR (with a comment or so) but maybe rather as tickets:

- What is the difference in treatment between battery and charging station? I compared the code for a few minutes, but I couldn't tell the major difference. Their docstring are also almost identical.
- Can we formally distinguish (check) that we treat a sensor/asset of correct type before we schedule?
- If the code is 75% (or so) identical, there might be an idea for future refactoring to note here.

Flix6x commented on flexmeasures/data/models/planning/battery.py:
```diff
@@ -37,6 +38,11 @@ def schedule_battery(
+    # Check for round-trip efficiency
+    if roundtrip_efficiency is None:
+        # Get default from sensor, or use 100% otherwise
+        roundtrip_efficiency = sensor.get_attribute("roundtrip_efficiency", 1)
```
Best to validate on input, but I have no objection raising here nonetheless. Note that we have no formal way yet for setting the `roundtrip_efficiency` attribute on a sensor.

Flix6x commented on flexmeasures/api/v1_3/routes.py:
```diff
@@ -104,6 +104,7 @@ def post_udi_event():
     This "PostUdiEventRequest" message posts a state of charge (soc) of 12.1 kWh at 10.00am,
     and a target state of charge of 25 kWh at 4.00pm,
     as UDI event 204 of device 10 of owner 7.
+    Roundtrip efficiency for use in scheduling is set to 98%.
```
No

Flix6x commented on flexmeasures/data/models/planning/charging_station.py:
```diff
@@ -95,6 +101,10 @@ def schedule_charging_station(
     else:
         device_constraints[0]["derivative max"] = sensor.get_attribute("capacity_in_mw")
 
+    # Apply round-trip efficiency evenly to charging and discharging
```
I don't expect so. Even when applying the efficiency to one side, there is still a need to split the power flow into up and down flows, so the number of variables wouldn't change.

Flix6x commented:
> And one fundamental question: It seems we are going from one `power` variable to `ems_power`, `device_power_down` and `device_power_up`. Does this mean that in principle (although it is usually isn't optimal) we could charge and discharge at the same time?

Correct. At first I added the constraint `device_power_up * device_power_down = 0`, but that results in a non-linear (quadratic) program. The same goes for the constraint `commitment_downwards_deviation * commitment_upwards_deviation = 0`. Given efficiencies between 0 than 1 and given convex commitment prices, I believe we can go without these constraints.

> For full understanding, I wanted to see how the battery and charging stations differ. I had a tough time really seeing how that is, as battery.py and charging_station.py seem to share a large amount of code.
> 
> My assumption: They both schedule only one sensor (one for a battery, one for a charging station) against SOC targets for lowest costs. The "EMS" is not being optimized, but it actually is the environment of the 1 device (e.g. the grid).
> 
> Is that correct? The comments are sometimes talking about "devices" (plural) so I'm unsure.

Our generic device scheduler (`def device_scheduler`) is able to handle an EMS with multiple devices, with constraints on the EMS level and on the device level, and market commitment on the EMS level. A typical example is a house with many devices. The two specific schedulers we coupled to the FlexMeasures API are for an EMS with a single device, i.e. a standalone Charge Point and a standalone battery system.

> A few smaller questions, maybe addressable in this PR (with a comment or so) but maybe rather as tickets:
> 
>     * What is the difference in treatment between battery and charging station? I compared the code for a few minutes, but I couldn't tell the major difference. Their docstring are also almost identical.

The differences are few, actually. Perhaps they have been dissipating as development progressed. I think at this point it's only in the min/max state of charge (which is assumed to be a known sensor property for a battery, but not so for a Charge Point, because it depends on the connected EV).

>     * Can we formally distinguish (check) that we treat a sensor/asset of correct type before we schedule?

Good idea. I made a card for [project 6](https://github.com/FlexMeasures/flexmeasures/projects/6).

>     * If the code is 75% (or so) identical, there might be an idea for future refactoring to note here.

Refactoring is a good idea. I added a card to [project 6](https://github.com/FlexMeasures/flexmeasures/projects/6).

nhoening reviewed (approved):
Approved.

I feel that some comments you made during our discussion in this PR are worthwhile.

This one could be helpful in the solvers' docstring:

> This generic device scheduler ( is able to handle an EMS with multiple devices, with constraints on the EMS level and on the device level, and market commitment on the EMS level. A typical example is a house with many devices. 

The following one seems like a crucial list of assumptions we rely on - maybe in a comment:

(when discussing why we not make sure that charging and discharging cannot happen at the same time)

> Given efficiencies between 0 than 1 and given convex commitment prices, I believe we can go without these constraints.

Are convex commitment prices a sure thing?
